src/i_oplmusic.c

```diff
@@ -1610,26 +1610,32 @@ static void I_OPL_StopSong(void *handle)
     OPL_Unlock();
 }
 
+static void I_OPL_ShutdownMusic(void);
+
 static void I_OPL_UnRegisterSong(void *handle)
 {
     if (!music_initialized)
     {
         return;
     }
 
+    I_OPL_ShutdownMusic();
+
     if (handle != NULL)
     {
         MIDI_FreeFile(handle);
     }
 }
 
+static boolean OPL_InitMusic(void);
+
 static void *I_OPL_RegisterSong(void *data, int len)
 {
     midi_file_t *result;
 
     if (!music_initialized)
     {
-        return NULL;
+        OPL_InitMusic();
```

Comment:
It's ugly, but it's fast and allow `Mix_HookMusic`. Any suggestions?

Reply:
You mean, when we change the midi player from anything else to OPL, right? What if we need to change between OPL and, e.g. SDL, because MAP01 has a MIDI song, MAP02 has MP3 and MAP03 has another MIDI song?

Reply:
> What if we need to change between OPL and, e.g. SDL, because MAP01 has a MIDI song, MAP02 has MP3 and MAP03 has another MIDI song?

Yes, which is why `Mix_RegisterEffect()` is problematic. I fixed this - shutdown OPL when an MP3 song is registered with `I_RegisterMusic()` in `i_sound.c`.